File: Algorithms/p111_Minimum_Depth_of_Binary_Tree/p111_Minimum_Depth_of_Binary_Tree_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if(!root) return 0;
        int depth = 0;
        deque<TreeNode*> q;
        q.push_back(root);
        while(!q.empty()){
            ++depth;
            int size = q.size();
            int i;
            for(i=0;i<size;++i){
                TreeNode* node = q.front();
                q.pop_front();
                if(!node->left&&!node->right) break;
                if(node->left) q.push_back(node->left);
                if(node->right) q.push_back(node->right);
            }
            if(i<size) break;
        }
        return depth;
    }
};

File: Algorithms/p111_Minimum_Depth_of_Binary_Tree/p111_Minimum_Depth_of_Binary_Tree_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        depth = 0
        q = collections.deque()
        q.append(root)
        while q:
            depth += 1
            done = False
            size = len(q)
            for i in range(size):
                node = q.popleft()
                if not node.left and not node.right:
                    done = True
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            if done:
                break
        return depth
